File: iosApp/iosApp/Theme/Localization.swift
import SwiftUI
import SharedLogic

private struct StringsKey: EnvironmentKey {
    static let defaultValue: any Strings = VietnameseStrings.shared
}

extension EnvironmentValues {
    /// The UI string table for the language the user picked in Settings.
    ///
    /// Read from the environment rather than the app state so a view that only
    /// shows text does not have to take the whole state object as a dependency.
    var strings: any Strings {
        get { self[StringsKey.self] }
        set { self[StringsKey.self] = newValue }
    }
}

extension View {
    func oddlyStrings(_ language: AppLanguage) -> some View {
        environment(\.strings, StringsCompanion.shared.of(language: language))
    }
}

extension LocalizedText {
    /// The side of this text matching the language currently on screen.
    func of(_ strings: any Strings) -> String {
        of(language: strings.language)
    }
}

extension Challenge {
    /// Resolves a challenge's content into the language currently on screen.
    func localized(_ strings: any Strings) -> LocalizedChallenge {
        localized(language: strings.language)
    }
}

extension Quote {
    /// Resolves a quote's content into the language currently on screen.
    func localized(_ strings: any Strings) -> LocalizedQuote {
        localized(language: strings.language)
    }
}

File: androidApp/src/main/kotlin/dev/lildua/oddly/ui/theme/Localization.kt
package dev.lildua.oddly.ui.theme

import androidx.compose.runtime.Composable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.staticCompositionLocalOf
import dev.lildua.oddly.core.text.Strings
import dev.lildua.oddly.core.text.VietnameseStrings
import dev.lildua.oddly.domain.model.Challenge
import dev.lildua.oddly.domain.model.LocalizedChallenge
import dev.lildua.oddly.domain.model.LocalizedQuote
import dev.lildua.oddly.domain.model.LocalizedText
import dev.lildua.oddly.domain.model.Quote

/**
 * The UI string table for the language the user picked in Settings.
 *
 * Static rather than dynamic because switching language re-composes the whole
 * tree anyway — there is no partial-update case worth optimising for.
 */
val LocalStrings = staticCompositionLocalOf<Strings> { VietnameseStrings }

/** Shorthand for the current language's side of a piece of content text. */
@Composable
@ReadOnlyComposable
fun LocalizedText.current(): String = of(LocalStrings.current.language)

/** Resolves a challenge's content into the language currently on screen. */
@Composable
@ReadOnlyComposable
fun Challenge.localized(): LocalizedChallenge = localized(LocalStrings.current.language)

/** Resolves a quote's content into the language currently on screen. */
@Composable
@ReadOnlyComposable
fun Quote.localized(): LocalizedQuote = localized(LocalStrings.current.language)
